sys/senml/phydat.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2021 Silke Hofstra
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include <stdint.h>
+
+#include "math.h"
+#include "senml.h"
+#include "senml/phydat.h"
+
+
+
+static uint8_t phydat_unit_to_senml_unit(uint8_t unit)
+{
+    switch (unit) {
+    /* Compatible units */
+    case UNIT_TEMP_C:  return SENML_UNIT_CELSIUS;
+    case UNIT_TEMP_K:  return SENML_UNIT_KELVIN;
+    case UNIT_LUX:     return SENML_UNIT_LUX;
+    case UNIT_M:       return SENML_UNIT_METER;
+    case UNIT_M2:      return SENML_UNIT_SQUARE_METER;
+    case UNIT_M3:      return SENML_UNIT_CUBIC_METER;
+    case UNIT_GR:      return SENML_UNIT_GRAM;
+    case UNIT_A:       return SENML_UNIT_AMPERE;
+    case UNIT_V:       return SENML_UNIT_VOLT;
+    case UNIT_W:       return SENML_UNIT_WATT;
+    case UNIT_COULOMB: return SENML_UNIT_COULOMB;
+    case UNIT_F:       return SENML_UNIT_FARAD;
+    case UNIT_OHM:     return SENML_UNIT_OHM;
+    case UNIT_PH:      return SENML_UNIT_PH;
+    case UNIT_PA:      return SENML_UNIT_PASCAL;
+    case UNIT_CD:      return SENML_UNIT_CANDELA;
+
+    /* Compatible Secondary units */
+    case UNIT_DBM:     return SENML_UNIT_DECIBEL_MILLIWATT;
+    case UNIT_PERCENT: return SENML_UNIT_PERCENT;
+    case UNIT_PERMILL: return SENML_UNIT_PERMILLE;
+    case UNIT_PPM:     return SENML_UNIT_PARTS_PER_MILLION;
+    case UNIT_PPB:     return SENML_UNIT_PARTS_PER_BILLION;
+
+    /* Incompatible units */
+    case UNIT_TEMP_F:  return SENML_UNIT_NONE;      /* use K or Cel instead */
+    case UNIT_GS:      return SENML_UNIT_NONE;      /* use T instead */
+    case UNIT_G:       return SENML_UNIT_NONE;      /* use m/s2 instead */
+    case UNIT_BAR:     return SENML_UNIT_NONE;      /* use Pa or hPa instead */
+    case UNIT_TIME:    return SENML_UNIT_NONE;      /* split into second/minute/hour */
+    case UNIT_DATE:    return SENML_UNIT_NONE;      /* split into day/month/year */
+    case UNIT_GPM3:    return SENML_UNIT_NONE;      /* use kg/m3 instead */
+    case UNIT_DPS:     return SENML_UNIT_NONE;      /* no alternative */
+    case UNIT_CPM3:    return SENML_UNIT_NONE;      /* no alternative */
```

Comment:
These two would make fine additions to the SenML registry (CPM3 as primary, and DPS as secondary to a to-be-added rad/s, see https://github.com/RIOT-OS/RIOT/issues/17571#issuecomment-1022161280). The registry is rather open, a mail to core@ietf.org has sufficed in the past. Would you care to just add them to SenML?

(Yes, this is a blunt attempt to get RIOT people more familiar with IETF workings ;-) )

Reply:
I think having these added to the SenML registry is a great idea, but maybe we should do that as part of #17571 so we can get this PR merged.

Reply:
Sure, doesn't need to be in this PR -- but anyway, the change would not happen until the appointed experts have approved the registry extension. Would you care to make the request?

Reply:
Sure, I have added a proposal to #17571 to allow others to comment before I submit them though.